### ISSUE TITLE
Double free corruption on ppc64le with XL

### ISSUE THREAD

**Issue by chuckatkins** [ornladios/ADIOS2#2144]
See https://open.cdash.org/testDetails.php?test=872605402&build=6476675

**Comment by KyleFromKitware** [comment]
Confirmed in my own testing. Investigating now.

**Comment by KyleFromKitware** [comment]
For my testing, I've been using the `Interface.ADIOSDefineAttributeTest.GetAttribute` test. Here's what I've found so far.

In this test, the failure is being caused by a second or invalid call to `free()` inside the `std::string` destructor for `m_TestData.S1array[0]`, inside `TEST_F(ADIOSDefineAttributeTest, GetAttribute)`. What's interesting is that when I put the following at the end of the test function:

```c++
    std::string& asdf = m_TestData.S1array[0];
    printf("%p %p %p %i\n", &asdf, &asdf[0], "one", sizeof(std::string));
```

I don't get the double-free/corruption error. It seems that referencing the string's `[0]` in a mutable fashion mutates the state of the string somehow, bringing it out of some ugly state. However, changing it to:

```c++
    const std::string& asdf = m_TestData.S1array[0];
    printf("%p %p %p %i\n", &asdf, &asdf[0], "one", sizeof(std::string));
```

causes the error to reappear, since the string is not being mutated.

**Comment by KyleFromKitware** [comment]
Unfortunately, I have not been able to find debuginfo for `libstdc++`, and have not been able to get a closer look at what's going on inside `std::string`.

**Comment by chuckatkins** [comment]
So the working theory is something wonky about how XL is doing `std::string` handling, which would in turn be pervasive across the rest of adios given the extent of string handling that's done.  Seems entirely plausible.

Do you think you could try to recreate a micro-example that reproduces the issue outside of adios so we can determine if it's a compiler bug or perhaps just some subtle undefined behavior we're relying on?

**Comment by KyleFromKitware** [comment]
That is going to be my next step. I took a brief crack at it yesterday but didn't have any luck. I'll try doing some things a little differently.

**Comment by KyleFromKitware** [comment]
OK, I have a minimal working example:

```c++
#include <array>
#include <string>

struct TestStruct
{
  std::array<std::string, 1> arr = {{"one"}};
};

int main(void)
{
  TestStruct s;
  return 0;
}
```

I will see if I can find some help in debugging this.

**Comment by KyleFromKitware** [comment]
Hmmm. I'm having a really hard time finding a place to report a bug to IBM.

I'll see if I can find a workaround to this issue.

**Comment by KyleFromKitware** [comment]
Interestingly, this works fine:

```c++
#include <array>
#include <string>

struct TestStruct
{
  std::array<std::string, 2> arr = {{"one", "two"}};
};

int main(void)
{
  TestStruct s;
  return 0;
}
```

**Comment by KyleFromKitware** [comment]
I'm still investigating the example above, but I'm starting to think that you have much bigger problems with the XL compiler. From what I can tell, the code that's being emitted by XL for this minimal example is not sane. It seems to be copying a `std::string` by copying the internal pointer, rather than the data, and then destroying the original `std::string`, thus `free()`ing the original pointer, which is still being held by the new `std::string`. Then, when the new `std::string` is destroyed, its internal pointer, which was already `free()`d, gets `free()`d, again, resulting in a double-free.

This seems like such a basic thing for a compiler to get wrong that I'm surprised XL works at all for this project.

**Comment by chuckatkins** [comment]
So it looks like XL is performing shallow copies when deep copies are expected?

**Comment by KyleFromKitware** [comment]
Something like that. It's a bit hard to make sense of the disassembly.

The fact that this only happens when a `std::string` is inside an `std::array` with exactly one element, which is in turn inside a struct, makes me wonder if there is some sort of bad premature optimization that XL is trying to do (even with `-O0`.)

I don't know any of this for sure... I'm just making best guesses with the information I have.

**Comment by KyleFromKitware** [comment]
I know that the `TestStruct` destructor, `std::array` destructor, and `std::string` destructor are all being called twice, on different objects. However, the `free()` that they boil down to tries to free the same pointer twice.

**Comment by chuckatkins** [comment]
@KyleFromKitware based on your detective work, it seems like a compiler bug in code generation for string arrays.  I've replaced those pieces in the testing with `std::vector` and it seems to have resolved the issue: https://github.com/ornladios/ADIOS2/pull/2126/commits/1772506892d2cbbb3acd4828539d1825996a626d

**Comment by KyleFromKitware** [comment]
@chuckatkins Thanks. However, locally I'm still seeing a number of other test failures that I'm still trying to investigate.

I see you put that commit in #2126. Are all the tests passing there?

**Comment by chuckatkins** [comment]
They are but I'm only running a small restricted set because of the slow
CPU emulation

**Comment by chuckatkins** [comment]
Fixed by #2126
